### PR TITLE
[ID-335] Remove account from the system - all BBs

### LINKED ISSUE #335: Remove account from the system - all BBs
I would like to bring up a discussion what we should do on back-end side when the user removes his account. 

As there was a requirement once the user clicks the "Remove" button it should remove the account like the user has never met the system.

Currently the account is removed from the Core BB but his corresponding data in the other building blocks is available.

I think we should open API in the Core BB which can be used (once a day for example) by the other building blocks to know which users have been recently deleted and to delete them in the their building block too.

@shurwit , @mdryankov , share your thoughts if you see any issues with this approach.

cc @pmarkhennessy

### PR DESCRIPTION
**Pull request by roberlander2**
## Description
This PR adds a `Deleted` flag to `OrgAppMembership` to mark app memberships as deleted as requested by a user. These deleted memberships will have all but their identifying data (`ID` and `AppOrgID`) removed and they will be retained until all interested building blocks have been informed of the update. The memberships will be deleted permanently only after this occurs.

**Edit (from the comments below):** The above statement relies on the assumption that all of the interested building blocks are working as expected. The interested building blocks are expected to make one request to the Core BB to get deleted account/membership information within the period between runs of the timer on the Core BB which finally deletes the marked account/membership data from the DB. The Core BB does not actually verify that all building blocks interested in deleted accounts/memberships have received the data before the timer performs the final deletion.

**Depends on #319**
**Resolves #335**

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [ ] Within a week
- [x] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Other (any another change that does not fall in one of the above categories.)

## Checklist:
_Please select all applicable options:_

[comment]: # (To select your options, please put an 'x' in the all boxes that apply.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [x] I have signed the Rokwire Contributor License Agreement ([CLA](https://rokwire.org/rokwire_cla)). (_Any contributor who is not an employee of the University of Illinois whose official duties include contributing to the Rokwire software, or who is not paid by the Rokwire project, needs to sign the CLA before their contribution can be accepted._)
- [x] I have updated the [CHANGELOG](../CHANGELOG.md).
- [x] I have read the [Contributor Guidelines](../CONTRIBUTING.md).
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My change requires updating the documentation.
- [x] I have made necessary changes to the documentation.
- [ ] I have added tests related to my changes.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

---

[comment]: # (Template credit: This pull request template is based on Embedded Artistry {https://github.com/embeddedartistry/templates/blob/master/.github/PULL_REQUEST_TEMPLATE.md}, Clowder {https://github.com/clowder-framework/clowder/blob/develop/.github/PULL_REQUEST_TEMPLATE.md}, and TalAter {https://github.com/TalAter/open-source-templates} templates.)

### CODE REVIEW

**Comment by petyos** [comment]
Hi @shurwit and @roberlander2, is this still valid from today point of view? Thanks

**Comment by shurwit** [comment]
> Hi @shurwit and @roberlander2, is this still valid from today point of view? Thanks

Hi @petyos, yes this will still be needed, thanks for checking. We will work on updating the PR when we get a chance, then re-request your review.

**Comment by roberlander2** [comment]
Hi @shurwit and @petyos, this PR is ready for review again.

**Review by petyos — COMMENTED** [review]
Hi @roberlander2 , thanks for your PR.

Some common comments:
As I see you have taken the approach to keep the deleted accounts within the original accounts structure together with the active accounts. This has disadvantages as this mixes them instead of keeping the deleted ones uncoupled from the active accounts in a separate collection which would serve only the deletion mechanism. As you have implemented it then I am ok with it, let's @shurwit  also confirm this is good in his opinion.

More specific questions:
As I understand when the user deletes account/membership it is only marked as "for deletion". After that a timer removes all data for the marked accounts except the one needed for the deleted-membeships API. Is this the process as a whole?

You also say "_..  removed and they will be retained until all interested building blocks have been informed of the update. The memberships will be deleted permanently only after this occurs._" How Does Core BB know when to fully remove all removed accounts and verify all interested BBs have removed all needed?

I also left one comment in the code. Thanks.

**Comment by roberlander2** [comment]
> Hi @roberlander2 , thanks for your PR.
> 
> Some common comments: As I see you have taken the approach to keep the deleted accounts within the original accounts structure together with the active accounts. This has disadvantages as this mixes them instead of keeping the deleted ones uncoupled from the active accounts in a separate collection which would serve only the deletion mechanism. As you have implemented it then I am ok with it, let's @shurwit also confirm this is good in his opinion.

Thanks @petyos, sorry for the delay in responding to you on this. @shurwit and I talked about your suggested separate deleted accounts and active accounts collections, and we agreed that it would be better than what I currently have implemented, so I will make this change and re-request your review.

> More specific questions: As I understand when the user deletes account/membership it is only marked as "for deletion". After that a timer removes all data for the marked accounts except the one needed for the deleted-membeships API. Is this the process as a whole?

When a user deletes his account/membership, all data associated with that account/membership is deleted except for information that is required to specify to other building blocks which account/membership was deleted. The timer performs the final deletion of the account document from the accounts collection or the membership sub-document from the account, depending on the situation.

> You also say "_.. removed and they will be retained until all interested building blocks have been informed of the update. The memberships will be deleted permanently only after this occurs._" How Does Core BB know when to fully remove all removed accounts and verify all interested BBs have removed all needed?

You are right, I should have written something different here. This statement relies on the assumption that all of the interested building blocks are working as expected. The interested building blocks are expected to make one request to the Core BB to get deleted account/membership information within the period between runs of the timer on the Core BB which finally deletes the marked account/membership data from the DB. The Core BB does not actually verify that all building blocks interested in deleted accounts/memberships have received the data before the timer performs the final deletion.

File: driven/storage/adapter.go [hunk 1]
```diff
@@ -614,6 +614,75 @@ func (sa *Adapter) getCachedApplicationOrganizations() ([]model.ApplicationOrgan
...
+		}
+
+		ids := make([]string, len(appOrgs))
+		for i, appOrg := range appOrgs {
+			ids[i] = appOrg.ID
+		}
+		return ids, nil
+	}
+
+	return nil, nil // nil slice for appID=authutils.AllApps, orgID=authutils.AllOrgs
```
[the remark this answers is not in the document]
Comment (roberlander2) [reply driven/storage/adapter.go:683]: Yes, this function will always return nil for our current service accounts. However, I am using this function to account for the other app_id, org_id pair cases in the situation that we later add a service account or have an interested third-party service that does not use "all", "all" when retrieving deleted account information.

**Comment by petyos** [comment]
Ok, Thanks!

**Comment by roberlander2** [comment]
Hi @petyos, this PR is ready for review again. Please let me know if you have any additional questions or concerns. Thanks!

**Comment by petyos** [comment]
Hi @roberlander2,

Thanks for reworking this. Keeping the deleted ones outside the accounts is much more clear.

How often have you designed the deleted ones to be removed from core BB? In my opinion we should not remove them or if we remove them this should happen very rarely - maybe once a week or if we want shorter then min once a day during the night. In this way we will be sure that there is no BB which has missed a removed account. In general the deleted ones' records will not be a huge number and we can allow them to live for a long time in the deleted memberships collection. As I see the default time you have set is 2 hours. What do you think about it?

File: driver/web/apis_services.go [hunk 2]
```diff
@@ -445,7 +445,17 @@ func (h ServicesApisHandler) deleteAccount(l *logs.Log, r *http.Request, claims
 		apps = append(apps, claims.AppID)
 	}
 
-	err := h.coreAPIs.Services.SerDeleteAccount(claims.Subject, apps)
+	var deleteContextBody []Def.DeletedMembershipContext
+	err := json.NewDecoder(r.Body).Decode(&deleteContextBody)
+	if err != nil {
+		return l.HTTPResponseErrorAction(logutils.ActionUnmarshal, "app membership delete context", nil, err, http.StatusBadRequest, true)
```
Comment (petyos) [inline driver/web/apis_services.go:451]: It fails with:
```
{
    "message": "Error unmarshalling app membership delete context: EOF",
    "status": "bad-request"
}
```
when we do not pass a body.

**Review by petyos — COMMENTED** [review]
Hi @roberlander2 , all this looks very well.

I've just added some minor issues in the review. Let me know if you have any questions or comments. Thanks.

**Comment by petyos** [comment]
Hi @roberlander2 I did small tweaks and I am merging it. Let me know if you have any comments or questions.

**Pull request merged**